### ISSUE TITLE
Add an "epoch"

### ISSUE THREAD

**Issue by pmetzger** [ocaml/opam#3537]
Dealing with packages whose upstream changed its numbering scheme is difficult in opam because it lacks a notion of "epoch". This is found in several other packaging schemes (like debian's dpkg, MacPorts, and others that I am familiar with.) The functionality in those systems works like this: the epoch is a counter in the package configuration file that can be incremented whenever a totally new upstream versioning scheme has been adopted; any package version whose epoch is larger than another's is understood to be newer than the other package.

I'd propose adding an optional "epoch" field to opam files. By default, any opam file without an epoch field would be considered to be at epoch 0, and most files should probably be left that way (without an explicit `epoch: 0`) . If an epoch field is added, it should be started at 1, and incremented by 1 if and only if the upstream maintainers change their numbering scheme.

Alternatively, epochs could be added to the directory name version number scheme, but that seems more complicated and would have to be done with great care.

**Comment by rjbou** [comment]
It is a good idea, and will help to maintain precedence after numbering scheme change. We will need to define what is such change as version comparison handles already some [changes](http://opam.ocaml.org/2.0-preview/doc/Manual.html#Package-Formulas) as it uses the [debian version comparison](https://github.com/ocaml/opam/blob/master/src/core/opamVersionCompare.mli#L23) (without checking that it complies to such format).

I'll keep in mind for 2.1.0 ;)

> Alternatively, epochs could be added to the directory name version number scheme, but that seems more complicated and would have to be done with great care.

Yes, this kind of change in the repo is more _difficult_ to perform. For information purpose, updating `opam admin` to handle it as a first step is more conceivable.

Thanks for the suggestion!

**Comment by paurkedal** [comment]
How do we deal with constraints from other packages to a past epoch of a package?  I found some docs related to [Python][1] and [Rust][2] after a quick search, which indicates quite different ways of dealing with it.  BTW, also rpm also has epochs.

[1]: https://legacy.python.org/dev/peps/pep-0440/#version-epochs
[2]: https://github.com/rust-lang/rfcs/blob/master/text/2052-epochs.md

**Comment by pmetzger** [comment]
You can treat an `epoch` as a new most significant field. If it's necessary to constrain the version number to an earlier epoch, you can add syntax to explicitly mention an optional epoch number as a leading field, though it will be important to be able to have the distinction between the epoch and "normal" part of the version number unambiguous, so a separator that doesn't appear in any real versions will be needed.

**Comment by paurkedal** [comment]
Yes, that's what the PEP-0440 does.  It uses `!` as separator between epoch and version, but both debian and rpm uses `:`, so maybe that's the more common.
